databricks_cli/stack/api.py

```diff
@@ -0,0 +1,336 @@
+        necessary fields for stack deployment to function well.
+
+        If there is an error here, then it is either an implementation error that must be fixed by
+        a developer or the User edited the stack status file created by the program.
+
+        :param stack_status: dict- stack status that is created by the program.
+        :return: None. Raises errors to stop deployment if there is a problem.
+        """
+        if STACK_NAME not in stack_status:
+            raise StackError("'%s' not in status" % STACK_NAME)
+        if STACK_RESOURCES not in stack_status:
+            raise StackError("'%s' not in status" % STACK_RESOURCES)
+        if STACK_DEPLOYED not in stack_status:
+            raise StackError("'%s' not in status" % STACK_DEPLOYED)
+        for deployed_resource in stack_status[STACK_DEPLOYED]:
+            if RESOURCE_ID not in deployed_resource:
+                raise StackError("%s doesn't exist in deployed resource status" % RESOURCE_ID)
+            if RESOURCE_SERVICE not in deployed_resource:
+                raise StackError("%s doesn't exist in deployed resource status" % RESOURCE_SERVICE)
+            if RESOURCE_PHYSICAL_ID not in deployed_resource:
+                raise StackError("%s doesn't exist in deployed resource status" %
+                                 RESOURCE_PHYSICAL_ID)
+
+    def validate_config(self, stack_config):
+        """
+        Validate fields within a stack configuration. This ensures that an inputted configuration
+        has the necessary fields for stack deployment to function well.
+
+        :param stack_config: dict- stack config that is inputted by the user.
+        :return: None. Raises errors to stop deployment if there is a problem.
+        """
+        if STACK_NAME not in stack_config:
+            raise StackError("'%s' not in configuration" % STACK_NAME)
+        if STACK_RESOURCES not in stack_config:
+            raise StackError("'%s' not in configuration" % STACK_RESOURCES)
+        for resource in stack_config[STACK_RESOURCES]:
+            if RESOURCE_ID not in resource:
+                raise StackError("%s doesn't exist in resource config" % RESOURCE_ID)
+            if RESOURCE_SERVICE not in resource:
+                raise StackError("%s doesn't exist in resource config" % RESOURCE_SERVICE)
+            if RESOURCE_PROPERTIES not in resource:
+                raise StackError("%s doesn't exist in resource config" % RESOURCE_PROPERTIES)
+
+    def deploy_config(self, stack_config, stack_status=None):
+        self.validate_config(stack_config)
+        if stack_status:
+            self.validate_status(stack_status)
+            resource_id_to_status = self._get_resource_to_status_map(stack_status)
+        else:
+            resource_id_to_status = {}
+
+        stack_name = stack_config[STACK_NAME]
+        click.echo('Deploying stack %s' % stack_name)
+        resource_statuses = []
+
+        for resource_config in stack_config[STACK_RESOURCES]:
```

Comment:
Don't we need to maintain a specific ordering of which resource is deployed first? If so, we need to enforce that here.

Reply:
So actually for jobs, workspaces and notebooks there doesn't need to be a specific order since we aren't running jobs from deployment. If we were to introduce clusters or job runs, then we would need to do an ordering.